File: problems/Validate Binary Search Tree/isValidBST.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public bool IsValidBST(TreeNode root) {
        // var inorderStore = new List<int>();
        // fillInorderStore(root, inorderStore);

        var inorderStore = getInorderStore(root);

        for (var i = 1; inorderStore.Count > i; ++i) {
            if (inorderStore[i] <= inorderStore[i - 1]) {
                return false;
            }
        }

        return true;
    }

    private void fillInorderStore(TreeNode root, List<int> store) {
        if (null == root) {
            return;
        }

        fillInorderStore(root.left, store);
        store.Add(root.val);
        fillInorderStore(root.right, store);
    }

    private List<int> getInorderStore(TreeNode root) {
        var store = new Stack<TreeNode>();
        var curr = root;
        var result = new List<int>();

        while (null != curr || 0 < store.Count) {
            while (null != curr) {
                store.Push(curr);
                curr = curr.left;
            }

            curr = store.Pop();
            result.Add(curr.val);
            curr = curr.right;
        }

        return result;
    }
}

File: problems/Validate Binary Search Tree/isValidBST.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        return dfs(root, null, null);
    }

    private boolean dfs(TreeNode node, Integer min, Integer max) {
        if (null == node) {
            return true;
        }

        if ((null != min && min >= node.val) || (null != max && max <= node.val)) {
            return false;
        }

        return dfs(node.left, min, node.val) && dfs(node.right, node.val, max);
    }
}
